README.md
# hydrusrv-sync [![Build status][travis-badge]][travis] [![Docker Hub build][docker-hub-badge]][docker-hub] [![Known vulnerabilities][snyk-badge]][snyk] [![JavaScript Standard Style][standardjs-badge]][standardjs]

> A tool to sync content from [hydrus server][hydrus-server] to
> [hydrusrv][hydrusrv]

__Adandoned: hydrusrv-sync is no longer maintained.__

hydrusrv-sync is a tool to sync content from [hydrus server][hydrus-server]'s
databases to [hydrusrv][hydrusrv]'s content database. It can be run manually
or automated (e.g., via cron job), depending on your needs.

## Table of contents

+ [Install](#install)
  + [Dependencies](#dependencies)
  + [Updating](#updating)
    + [Upgrading from 2.x to 3.x](#upgrading-from-2x-to-3x)
    + [Upgrading from 1.x to 2.x](#upgrading-from-1x-to-2x)
+ [Usage](#usage)
  + [Configuration](#configuration)
  + [Run](#run)
    + [Running with Docker](#running-with-docker)
+ [Caveats](#caveats)
+ [Maintainer](#maintainer)
+ [Contribute](#contribute)
+ [License](#license)

## Install

The easiest way to install is via cloning this repository:

```zsh
user@local:~$ git clone https://github.com/mserajnik/hydrusrv-sync.git
user@local:~$ cd hydrusrv-sync
user@local:hydrusrv-sync$ yarn
```

If you encounter any errors during installation, those are likely caused by
packages hydrusrv uses that utilize native bindings (e.g.,
[better-sqlite3][better-sqlite3]).

Please check the error message and contact the maintainers of those packages
directly if you cannnot resolve your issues.

### Dependencies

+ [hydrusrv][hydrusrv] (`6.x` or `7.x` for hydrusrv-sync `3.x`)
+ [hydrus server][hydrus-server] (installing and running the server is quite
  difficult and not recommended for people who have no prior experience with
  hydrus; see [here][hydrus-server-installation] for installation instructions)
+ [Node.js][node-js]
+ [Yarn][yarn]

I usually use the latest versions of Node.js and Yarn; if there has not been an
update to this repository in a while and something breaks on the latest
Node/Yarn version, please [let me know][issues].

### Updating

If you have installed via cloning the repository, you can update via Git:

```zsh
user@local:hydrusrv-sync$ git pull
user@local:hydrusrv-sync$ yarn
```

Always make sure to run `yarn` after updating to install any packages you might
be missing.

hydrusrv-sync follows [semantic versioning][semantic-versioning] and any
breaking changes that require additional attention will be released under a new
major version (e.g., `2.0.0`). Minor version updates (e.g., `1.1.0` or `1.2.0`)
are therefore always safe to simply install via the routine mentioned before.

When necessary, this section will be expanded with upgrade guides to new major
versions.

#### Upgrading from 2.x to 3.x

Upgrading from `2.x` to `3.x` can be done via `git pull && yarn`.

The major version bump was made due to the incompatibility of `2.x` with
hydrusrv `6.x` and `7.x`. So if you are using one of those versions, you need
to upgrade hydrusrv-sync to `3.x`.

#### Upgrading from 1.x to 2.x

Upgrading from `1.x` to `2.x` can be done via `git pull && yarn`.

The major version bump was made due to the incompatibility of `1.x` with
hydrusrv `5.x`. So if you are using that version, you need to upgrade
hydrusrv-sync to `2.x`.

## Usage

### Configuration

Duplicate `.env.example` as `.env`. Make use of the following options to
configure your installation:

+ `HYDRUSRV_CONTENT_DB_PATH=`: sets the path to hydrusrv's content database.
  __Absolute path required.__
+ `HYDRUS_SERVER_DB_PATH=`: sets the path to the hydrus server main database
  (called `server.db`). __Absolute path required.__
+ `HYDRUS_MASTER_DB_PATH=`: sets the path to the hydrus server master database
  (called `server.master.db`). __Absolute path required.__
+ `HYDRUS_MAPPINGS_DB_PATH=`: sets the path to the hydrus server mappings
  database (called `server.mappings.db`). __Absolute path required.__
+ `HYDRUS_TAG_REPOSITORY=2`: the ID of the hydrus server tag repository
  hydrusrv-sync should use.
+ `HYDRUS_FILE_REPOSITORY=3`: the ID of the hydrus server file repository
  hydrusrv-sync should use.
+ `HYDRUS_SUPPORTED_MIME_TYPES=1,2,3,4,9,14,18,20,21,23,25,26,27`: the IDs of
  the MIME types hydrusrv-sync should support. See [here][supported-mime-types]
  for the complete list of MIME types you can choose from.

### Run

You can run hydrusrv-sync manually:

```zsh
user@local:hydrusrv-sync$ ./bin/sync
```

hydrusrv-sync will output information about how long the different steps take
and how many files, tags and mappings are synced.

Generally, a sync should be really fast (a few seconds) on small databases with
only a few thousand files/tags. It will slow down considerably on large
databases, but if you only intend to sync once or twice a day, this should not
become an issue unless we are talking about tens of millions of files/tags.

Here is an example of a sync with a fairly large database:

```
12/20/2018, 5:18:25 PM: running sync...

create initial tables (if necessary): 0.003s
drop zombie tables: 0.000s
attach hydrus databases: 0.002s
get namespaces: 5.779s
create new tables: 0.002s
fill new namespaces table: 0.001s
fill new tags table: 18.410s
fill new files table: 17.038s
fill new mappings table: 32.535s
fill new mime types table: 0.046s
update tag counts on new files table: 4.646s
detach hydrus databases: 0.016s
replace current tables: 6.470s
clean up: 4.896s
total: 89.855s

namespaces: 15, tag: 73240, files: 216822, mappings: 7655592
```

hydrusrv-sync always copies over all the data without comparing changes. This
is because due to the nature of how hydrus server stores its data, running
comparisons would have been very unefficient and slow.

Drive longevity should generally not become an issue since you would have to
run syncs with enormous databases very frequently. E.g., even a hydrusrv
content dabase that contains a million files should typically only be around
1-3 GB in size, depending on the amount of tags and mappings. Running a sync
that fills this database in shorter intervals (let us say every two hours) will
only result in 12-36 GB written every day.

To quote [Ontrack's article on SSD lifetime][ontrack-ssd-lifetime]:

> "A typical TBW figure for a 250 GB SSD lies between 60 and 150 terabytes
> written. That means: To get over a guaranteed TBW of 70, a user would have to
> write 190(!) GB daily over a period of one year (In other words, to fill two
> thirds of the SSD with new data every day)."

However, if you are still worried and have enough RAM available, you can also
put the hydrusrv content database on a [tmpfs][tmpfs] or any other kind of
RAM-based storage; since all the data is copied over with every sync anyway
there is no actual need to persist the data between reboots.

Depending on your use case, setting up a cron job or using other tools to run
hydrusrv-sync automatically might make more sense that doing so manually.

When hydrusrv-sync runs it creates a `.sync-lock` file. This file is used to
determine if a sync is already in progress and hydrusrv-sync will exit early
if that is the case. Therefore, trying to run it in very short intervals will
usually not cause any issues, even if a previous sync has not yet been
completed.

The `.sync-lock` file will be deleted if a sync successfully completes or on
error, but if something should unexpectedly cause this not to happen, you can
of course also delete it manually.

### Running with Docker

You can also run hydrusrv-sync with [Docker][docker] if you want to. A
[Docker image][docker-hub] is available on Docker Hub. The image is set up
to run hydrusrv-sync once on container startup and via a configurable cron job
afterwards.

See [here][hydrusrv-docker] for a [Docker Compose][docker-compose] setup that
combines hydrus server, hydrusrv and hydrusrv-sync into an easy to use package.

If you want to create your own setup, please take a look at the
[Dockerfile](Dockerfile) and the
[entrypoint file](.docker/docker-entrypoint.sh) to figure out how to configure
it.

## Caveats

hydrusrv was mainly developed for my personal use and might therefore lack some
features others might want to see. Some of these could be:

+ hydrusrv-sync __needs__ one tag and one file repository to work. Trying to
  run it without either will result in errors. It also cannot support
  additional repositories.
+ hydrusrv-sync discards namespaces containing other characters than
  alphanumerics and underscores to not falsely assume emote tags like `>:)`
  have a namespace and to prevent errors (namespaces are used as column names
  and SQLite does not allow most characters aside from alphanumerics in those;
  mapping such special characters would have been possible, but did not seem
  worth the effort).
+ hydrus server supports many more MIME types than the ones I have limited
  hydrusrv-sync to. It only supports major image and video formats hydrusrv is
  supposed to serve.

In addition, you might run into issues or limitations when using hydrusrv-sync.
Here are the ones I am currently aware of:

+ When hydrusrv-sync copies the hydrus server data, it does so without locking
  the database to prevent issues in hydrus server (which most likely does not
  expect another application to randomly lock the database when it is trying to
  write). The update can also take a while on larger databases, which might
  lead to hydrus server changing or adding data while hydrusrv-sync is still
  running. When this happens, hydrusrv-sync will leave the existing data in
  hydrusrv's content database as it is and you will have to run it again.
+ hydrus client/server is updated frequently (usually once a week) and while I
  try to keep hydrusrv-sync up-to-date with any database changes (that
  thankfully do not occur very frequently), I suggest keeping an old copy of
  hydrus server when updating in case anything breaks. In addition, please
  [let me know][issues] if that happens.

## Maintainer

[mserajnik][maintainer]

## Contribute

You are welcome to help out!

[Open an issue][issues] or submit a pull request.

## License

[MIT](LICENSE.md) © Michael Serajnik

[travis]: https://travis-ci.com/mserajnik/hydrusrv-sync
[travis-badge]: https://travis-ci.com/mserajnik/hydrusrv-sync.svg

[docker-hub-badge]: https://img.shields.io/docker/automated/mserajnik/hydrusrv-sync.svg

[snyk]: https://snyk.io/test/github/mserajnik/hydrusrv-sync
[snyk-badge]: https://snyk.io/test/github/mserajnik/hydrusrv-sync/badge.svg

[standardjs]: https://standardjs.com
[standardjs-badge]: https://img.shields.io/badge/code_style-standard-brightgreen.svg

[hydrus-server]: http://hydrusnetwork.github.io/hydrus
[hydrusrv]: https://github.com/mserajnik/hydrusrv
[better-sqlite3]: https://github.com/JoshuaWise/better-sqlite3/wiki/Troubleshooting-installation
[hydrus-server-installation]: http://hydrusnetwork.github.io/hydrus/help/server.html
[node-js]: https://nodejs.org/en/
[yarn]: https://yarnpkg.com/
[semantic-versioning]: https://semver.org/
[ontrack-ssd-lifetime]: https://www.ontrack.com/blog/2018/02/07/how-long-do-ssds-really-last/
[tmpfs]: http://man7.org/linux/man-pages/man5/tmpfs.5.html
[docker]: https://www.docker.com/
[docker-hub]: https://hub.docker.com/r/mserajnik/hydrusrv-sync/
[hydrusrv-docker]: https://github.com/mserajnik/hydrusrv-docker
[docker-compose]: https://docs.docker.com/compose/
[supported-mime-types]: https://github.com/mserajnik/hydrusrv-sync/blob/master/src/config/index.js#L5-L17

[maintainer]: https://github.com/mserajnik
[issues]: https://github.com/mserajnik/hydrusrv-sync/issues/new


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.4.0] - 2018-03-11

### Changed

+ Updated dependencies

### Removed

+ Removed FOSSA badge

## [3.3.0] - 2018-02-15

### Changed

+ Updated dependencies
+ Excluded `.git` directory from Docker image

## [3.2.0] - 2018-01-29

### Changed

+ Updated dependencies

## [3.1.0] - 2018-01-15

### Changed

+ Updated dependencies

## [3.0.0] - 2018-12-28

### Added

+ Added `tag_count` field to `files` table
+ Added `mime_types` table

### Changed

+ Cleaned up code
+ Updated dependencies

### Fixed

+ Fixed wrong exit code when erroring on cleanup

## [2.0.0] - 2018-12-09

### Added

+ Added handling for cached file and tag count tables

### Changed

+ Updated dependencies

## [1.4.1] - 2018-12-05

### Fixed

+ Fixed wrong link in readme

## [1.4.0] - 2018-12-05

### Changed

+ Cleaned up environment variables
+ Updated dependencies

## [1.3.0] - 2018-12-05

### Changed

+ Moved content database checkpointing here from hydrusrv

## [1.2.0] - 2018-12-03

### Changed

+ Added indexes on `mappings` fields for increased query performance in
  hydrusrv
+ Dropped `hydrusrv` prefix from tables to be consistent with the hydrusrv
  application database naming scheme
+ File hashes in hex form are now computed on sync via SQLite's `HEX()`
  function (in combination with `LOWER()`) for increased performance in
  hydrusrv
+ Updated dependencies

## [1.1.0] - 2018-12-02

### Changed

+ Reduced Docker image size

### Fixed

+ Fixed letter case in readme

## 1.0.0 - 2018-12-01

### Added

+ Initial release

[Unreleased]: https://github.com/mserajnik/hydrusrv-sync/compare/3.4.0...develop
[3.4.0]: https://github.com/mserajnik/hydrusrv-sync/compare/3.3.0...3.4.0
[3.3.0]: https://github.com/mserajnik/hydrusrv-sync/compare/3.2.0...3.3.0
[3.2.0]: https://github.com/mserajnik/hydrusrv-sync/compare/3.1.0...3.2.0
[3.1.0]: https://github.com/mserajnik/hydrusrv-sync/compare/3.0.0...3.1.0
[3.0.0]: https://github.com/mserajnik/hydrusrv-sync/compare/2.0.0...3.0.0
[2.0.0]: https://github.com/mserajnik/hydrusrv-sync/compare/1.4.1...2.0.0
[1.4.1]: https://github.com/mserajnik/hydrusrv-sync/compare/1.4.0...1.4.1
[1.4.0]: https://github.com/mserajnik/hydrusrv-sync/compare/1.3.0...1.4.0
[1.3.0]: https://github.com/mserajnik/hydrusrv-sync/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/mserajnik/hydrusrv-sync/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/mserajnik/hydrusrv-sync/compare/1.0.0...1.1.0


tests/sync.js
const path = require('path')

const test = require('ava')
const fse = require('fs-extra')

const setup = require('./_setup')

setup.setTestEnvironment()

let app, db

test.before(t => {
  fse.copySync(
    path.resolve(__dirname, 'storage/content.db.template'),
    path.resolve(__dirname, `storage/content.db`)
  )

  app = require('../app')
  db = require('../src/db')

  db.connect()
})

test.serial('sync: run', t => {
  app.runSync()

  db.connect()

  t.deepEqual(
    db.hydrusrv.prepare(
      `SELECT COUNT(*) FROM namespaces
        UNION
      SELECT COUNT(*) FROM tags
        UNION
      SELECT COUNT(*) FROM files
        UNION
      SELECT COUNT(*) FROM mappings`
    ).pluck().all(),
    [1, 5, 10, 20]
  )
})

test.after.always(t => {
  db.close()

  fse.removeSync(path.resolve(__dirname, `storage/content.db`))
  fse.removeSync(path.resolve(__dirname, `storage/content.db-shm`))
  fse.removeSync(path.resolve(__dirname, `storage/content.db-wal`))
})


tests/_setup.js
const path = require('path')

module.exports = {
  setTestEnvironment () {
    process.env.NODE_ENV = 'test'
    process.env.HYDRUSRV_CONTENT_DB_PATH = path.resolve(
      __dirname, 'storage/content.db'
    )
    process.env.HYDRUS_SERVER_DB_PATH = path.resolve(
      __dirname, 'hydrus-server-dummy/server.db'
    )
    process.env.HYDRUS_MASTER_DB_PATH = path.resolve(
      __dirname, 'hydrus-server-dummy/server.master.db'
    )
    process.env.HYDRUS_MAPPINGS_DB_PATH = path.resolve(
      __dirname, 'hydrus-server-dummy/server.mappings.db'
    )
    process.env.HYDRUS_TAG_REPOSITORY = 2
    process.env.HYDRUS_FILE_REPOSITORY = 3
    process.env.HYDRUS_SUPPORTED_MIME_TYPES = '1,2,3,4,14,21,23'
  }
}


app.js
const config = require('./src/config')
const db = require('./src/db')
const profiler = require('./src/util/profiler')

module.exports = {
  runSync () {
    profiler.init()

    profiler.log('{datetime}: running sync...\n\n')

    db.connect()

    this.createTables(true)
    profiler.log('create initial tables (if necessary): {dt}\n')

    this.dropZombieTables()
    profiler.log('drop zombie tables: {dt}\n')

    db.attachHydrusDatabases()
    profiler.log('attach hydrus databases: {dt}\n')

    const namespaces = this.getNamespaces()
    profiler.log('get namespaces: {dt}\n')

    this.createTables(false, namespaces)
    profiler.log('create new tables: {dt}\n')

    this.fillNewNamespacesTable(namespaces)
    profiler.log('fill new namespaces table: {dt}\n')

    this.fillNewTagsTable()
    profiler.log('fill new tags table: {dt}\n')

    this.fillNewFilesTable(namespaces)
    profiler.log('fill new files table: {dt}\n')

    this.fillNewMappingsTable()
    profiler.log('fill new mappings table: {dt}\n')

    this.fillNewMimeTypesTable()
    profiler.log('fill new mime types table: {dt}\n')

    this.updateTagCountsOnNewFilesTables()
    profiler.log('update tag counts on new files table: {dt}\n')

    db.detachHydrusDatabases()
    profiler.log('detach hydrus databases: {dt}\n')

    this.replaceCurrentTables()
    profiler.log('replace current tables: {dt}\n')

    this.cleanUp()
    profiler.log('clean up: {dt}\n')

    profiler.log(`total: {t}\n\n`)

    console.info(this.getTotals())

    db.close()
  },
  abortSync () {
    process.nextTick(() => {
      db.close()
    })
  },
  createTables (initial = true, namespaces) {
    const suffix = initial ? '' : '_new'

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS namespaces${suffix} (
        id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT UNIQUE,
        name TEXT NOT NULL UNIQUE
      )`
    ).run()

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS tags${suffix} (
        id INTEGER NOT NULL PRIMARY KEY UNIQUE,
        name TEXT NOT NULL UNIQUE,
        file_count INTEGER NOT NULL,
        random TEXT NOT NULL
      )`
    ).run()

    const namespaceColumns = []

    if (Array.isArray(namespaces)) {
      for (const namespace of namespaces) {
        namespaceColumns.push(
          `namespace_${namespace.split(' ').join('_')} TEXT`
        )
      }
    }

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS files${suffix} (
        id INTEGER NOT NULL PRIMARY KEY UNIQUE,
        tags_id INTEGER UNIQUE DEFAULT NULL,
        hash TEXT UNIQUE NOT NULL,
        mime INTEGER NOT NULL,
        size INTEGER NOT NULL,
        width INTEGER NOT NULL,
        height INTEGER NOT NULL,
        tag_count INTEGER NOT NULL DEFAULT 0,
        random TEXT NOT NULL
        ${namespaceColumns.length ? ',' + namespaceColumns.join(',') : ''}
      )`
    ).run()

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS mappings${suffix} (
        file_tags_id INTEGER NOT NULL,
        tag_id INTEGER NOT NULL,
        FOREIGN KEY(file_tags_id) REFERENCES files${suffix}(tags_id)
          ON UPDATE CASCADE
          ON DELETE CASCADE,
        FOREIGN KEY(tag_id) REFERENCES tags${suffix}(id)
          ON UPDATE CASCADE
          ON DELETE CASCADE
      )`
    ).run()

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS mime_types${suffix} (
        id INTEGER UNIQUE NOT NULL PRIMARY KEY
      )`
    ).run()

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS tag_counts${suffix} (
        hash TEXT UNIQUE NOT NULL PRIMARY KEY,
        count INTEGER NOT NULL
      )`
    ).run()

    db.hydrusrv.prepare(
      `CREATE TABLE IF NOT EXISTS file_counts${suffix} (
        hash TEXT UNIQUE NOT NULL PRIMARY KEY,
        count INTEGER NOT NULL
      )`
    ).run()
  },
  dropZombieTables () {
    db.hydrusrv.prepare('DROP TABLE IF EXISTS namespaces_new').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS mappings_new').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS tags_new').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS files_new').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS mime_types_new').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS tag_counts_new').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS file_counts_new').run()
  },
  getNamespaces () {
    return db.hydrusrv.prepare(
      `SELECT name FROM (
        SELECT DISTINCT SUBSTR(
          ${config.hydrusTableTags}.tag,
          0,
          INSTR(${config.hydrusTableTags}.tag, ':')
        ) AS name
        FROM
          ${config.hydrusTableCurrentMappings}
        NATURAL JOIN
          ${config.hydrusTableRepositoryTagIdMap}
        NATURAL JOIN
          ${config.hydrusTableTags}
        NATURAL JOIN
          ${config.hydrusTableRepositoryHashIdMapTags}
        NATURAL JOIN
          ${config.hydrusTableFilesInfo}
        WHERE
          ${config.hydrusTableTags}.tag LIKE '%_:_%'
        AND
          ${config.hydrusTableFilesInfo}.mime IN (
            ${config.supportedMimeTypes}
          )
      )
      WHERE
        name REGEXP '^[a-zA-Z0-9_]+$'
      ORDER BY
        name`
    ).pluck().all()
  },
  fillNewNamespacesTable (namespaces) {
    for (const namespace of namespaces) {
      db.hydrusrv.prepare(
        'INSERT INTO namespaces_new (name) VALUES (?)'
      ).run(namespace)
    }
  },
  fillNewTagsTable () {
    db.hydrusrv.prepare(
      `INSERT INTO tags_new
        SELECT
          ${config.hydrusTableCurrentMappings}.service_tag_id,
          ${config.hydrusTableTags}.tag,
          COUNT(*),
          SUBSTR(''||RANDOM(), -4)
        FROM
          ${config.hydrusTableCurrentMappings}
        NATURAL JOIN
          ${config.hydrusTableRepositoryTagIdMap}
        NATURAL JOIN
          ${config.hydrusTableTags}
        NATURAL JOIN
          ${config.hydrusTableRepositoryHashIdMapTags}
        NATURAL JOIN
          ${config.hydrusTableFilesInfo}
        WHERE
          ${config.hydrusTableFilesInfo}.mime IN (
            ${config.supportedMimeTypes}
          )
        GROUP BY
          ${config.hydrusTableTags}.tag`
    ).run()
  },
  fillNewFilesTable (namespaces) {
    const namespaceColumns = []

    for (const namespace of namespaces) {
      namespaceColumns.push(
        `namespace_${namespace.split(' ').join('_')}`
      )
    }

    db.hydrusrv.prepare(
      `INSERT INTO files_new (
        id,
        tags_id,
        hash,
        mime,
        size,
        width,
        height,
        random
        ${namespaceColumns.length ? ',' + namespaceColumns.join(',') : ''}
      )
        SELECT
          ${config.hydrusTableCurrentFiles}.service_hash_id,
          ${config.hydrusTableRepositoryHashIdMapTags}.service_hash_id,
          LOWER(HEX(${config.hydrusTableHashes}.hash)),
          ${config.hydrusTableFilesInfo}.mime,
          ${config.hydrusTableFilesInfo}.size,
          ${config.hydrusTableFilesInfo}.width,
          ${config.hydrusTableFilesInfo}.height,
          SUBSTR(''||random(), -4)
          ${namespaceColumns.length ? ', null AS ' + namespaceColumns.join(', null AS ') : ''}
        FROM
          ${config.hydrusTableHashes}
        NATURAL JOIN
          ${config.hydrusTableFilesInfo}
        NATURAL JOIN
          ${config.hydrusTableRepositoryHashIdMapFiles}
        LEFT JOIN
          ${config.hydrusTableRepositoryHashIdMapTags}
          ON ${config.hydrusTableRepositoryHashIdMapTags}.master_hash_id =
            ${config.hydrusTableHashes}.master_hash_id
        NATURAL JOIN
          ${config.hydrusTableCurrentFiles}
        WHERE
          ${config.hydrusTableFilesInfo}.mime IN (
            ${config.supportedMimeTypes}
          )`
    ).run()

    db.hydrusrv.prepare(
      `CREATE TEMP TABLE temp_namespaces_reduced AS
        SELECT
          master_tag_id, tag
        FROM
          ${config.hydrusTableTags}
        WHERE
          tag LIKE '%_:_%'`
    ).run()

    const selectStatement = db.hydrusrv.prepare(
      `SELECT
        REPLACE(temp_namespaces_reduced.tag, :namespace, '') AS tag,
        ${config.hydrusTableRepositoryHashIdMapTags}.service_hash_id AS tags_id
      FROM
        ${config.hydrusTableCurrentMappings}
      NATURAL JOIN
        ${config.hydrusTableRepositoryTagIdMap}
      NATURAL JOIN
        temp_namespaces_reduced
      NATURAL JOIN
        ${config.hydrusTableRepositoryHashIdMapTags}
      WHERE
        temp_namespaces_reduced.tag LIKE :namespace || '_%'
      GROUP BY
        tags_id`
    )

    const updateStatements = []

    namespaces.map((namespace, i) => {
      updateStatements[namespaces[i]] = db.hydrusrv.prepare(
        `UPDATE
          files_new
        SET
          namespace_${namespace.replace(' ', '_')} = :tag
        WHERE
          tags_id = :tags_id`
      )
    })

    db.hydrusrv.transaction(namespaces => {
      for (const namespace of namespaces) {
        const tags = selectStatement.all({
          namespace: `${namespace}:`
        })

        db.hydrusrv.transaction(tags => {
          for (const tag of tags) {
            updateStatements[namespace].run(tag)
          }
        })(tags)
      }
    })(namespaces)

    db.hydrusrv.prepare('DROP TABLE temp_namespaces_reduced').run()
  },
  fillNewMappingsTable () {
    db.hydrusrv.prepare(
      `INSERT INTO mappings_new
        SELECT
          ${config.hydrusTableCurrentMappings}.service_hash_id,
          ${config.hydrusTableCurrentMappings}.service_tag_id
        FROM
          ${config.hydrusTableCurrentMappings}
        NATURAL JOIN
          ${config.hydrusTableRepositoryHashIdMapTags}
        NATURAL JOIN
          ${config.hydrusTableFilesInfo}
        INNER JOIN
          ${config.hydrusTableRepositoryHashIdMapFiles}
          ON ${config.hydrusTableRepositoryHashIdMapFiles}.master_hash_id =
            ${config.hydrusTableFilesInfo}.master_hash_id
        INNER JOIN
          ${config.hydrusTableCurrentFiles}
          ON ${config.hydrusTableCurrentFiles}.service_hash_id =
            ${config.hydrusTableRepositoryHashIdMapFiles}.service_hash_id
        WHERE
          ${config.hydrusTableFilesInfo}.mime IN (
            ${config.supportedMimeTypes}
          )`
    ).run()
  },
  fillNewMimeTypesTable () {
    db.hydrusrv.prepare(
      `INSERT INTO mime_types_new
        SELECT DISTINCT
          mime
        FROM
          files_new`
    ).run()
  },
  updateTagCountsOnNewFilesTables () {
    db.hydrusrv.prepare(
      `CREATE INDEX
        temp_idx_mappings_file_tags_id
      ON
        mappings_new(file_tags_id)`
    ).run()

    db.hydrusrv.prepare(
      `UPDATE
        files_new
      SET
        tag_count = (
          SELECT
            COUNT(*)
          FROM
            mappings_new
          WHERE
            mappings_new.file_tags_id = files_new.tags_id
        )`
    ).run()

    db.hydrusrv.prepare('DROP INDEX temp_idx_mappings_file_tags_id').run()
  },
  replaceCurrentTables () {
    db.hydrusrv.prepare('DROP TABLE IF EXISTS namespaces').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS mappings').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS tags').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS files').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS mime_types').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS tag_counts').run()
    db.hydrusrv.prepare('DROP TABLE IF EXISTS file_counts').run()

    db.hydrusrv.prepare('ALTER TABLE namespaces_new RENAME TO namespaces').run()
    db.hydrusrv.prepare('ALTER TABLE tags_new RENAME TO tags').run()
    db.hydrusrv.prepare('ALTER TABLE files_new RENAME TO files').run()
    db.hydrusrv.prepare('ALTER TABLE mappings_new RENAME TO mappings').run()
    db.hydrusrv.prepare('ALTER TABLE mime_types_new RENAME TO mime_types').run()
    db.hydrusrv.prepare('ALTER TABLE tag_counts_new RENAME TO tag_counts').run()
    db.hydrusrv.prepare('ALTER TABLE file_counts_new RENAME TO file_counts').run()

    db.hydrusrv.prepare(
      'CREATE INDEX idx_mappings_file_tags_id ON mappings(file_tags_id)'
    ).run()
    db.hydrusrv.prepare(
      'CREATE INDEX idx_mappings_tag_id ON mappings(tag_id)'
    ).run()
  },
  getTotals () {
    return db.hydrusrv.prepare(
      `SELECT COUNT(*) FROM namespaces
        UNION
      SELECT COUNT(*) FROM tags
        UNION
      SELECT COUNT(*) FROM files
        UNION
      SELECT COUNT(*) FROM mappings`
    ).pluck().all().map(
      (count, i) => ['namespaces: ', 'tag: ', 'files: ', 'mappings: '][i] +
        count
    ).join(', ')
  },
  cleanUp () {
    try {
      db.hydrusrv.prepare('VACUUM').run()
      db.hydrusrv.pragma('wal_checkpoint(TRUNCATE)')
    } catch (err) {
      console.info(
        'could not clean up after succesful sync, will try again on the next' +
          'run.'
      )
    }
  }
}


src/db/index.js
const Database = require('better-sqlite3')

const config = require('../config')

module.exports = {
  connect () {
    this.hydrusrv = new Database(config.hydrusrvContentDbPath, {
      fileMustExist: true
    })

    this.setWalMode()
    this.addFunctions()
  },
  close () {
    this.hydrusrv.close()
  },
  setWalMode () {
    this.hydrusrv.pragma('journal_mode = WAL')
  },
  addFunctions () {
    this.hydrusrv.function(
      'regexp', (pattern, string) => {
        if (pattern && string) {
          return string.match(new RegExp(pattern)) !== null ? 1 : 0
        }

        return null
      }
    )
  },
  attachHydrusDatabases () {
    this.hydrusrv.prepare(
      `ATTACH '${config.hydrusServerDbPath}' AS hydrus_server_db`
    ).run()
    this.hydrusrv.prepare(
      `ATTACH '${config.hydrusMasterDbPath}' AS hydrus_master_db`
    ).run()
    this.hydrusrv.prepare(
      `ATTACH '${config.hydrusMappingsDbPath}' AS hydrus_mappings_db`
    ).run()
  },
  detachHydrusDatabases () {
    this.hydrusrv.prepare('DETACH hydrus_server_db').run()
    this.hydrusrv.prepare('DETACH hydrus_master_db').run()
    this.hydrusrv.prepare('DETACH hydrus_mappings_db').run()
  }
}


.docker/docker-entrypoint.sh
#!/bin/sh

cp .env.example .env

HYDRUSRV_CONTENT_DB_PATH="${HYDRUSRV_CONTENT_DB_PATH:=/data/content.db}"
HYDRUS_SERVER_DB_PATH="${HYDRUS_SERVER_DB_PATH:=/data/server.db}"
HYDRUS_MASTER_DB_PATH="${HYDRUS_MASTER_DB_PATH:=/data/server.master.db}"
HYDRUS_MAPPINGS_DB_PATH="${HYDRUS_MAPPINGS_DB_PATH:=/data/server.mappings.db}"
HYDRUS_TAG_REPOSITORY="${HYDRUS_TAG_REPOSITORY:=2}"
HYDRUS_FILE_REPOSITORY="${HYDRUS_FILE_REPOSITORY:=3}"
HYDRUS_SUPPORTED_MIME_TYPES="${HYDRUS_SUPPORTED_MIME_TYPES:=1,2,3,4,9,14,18,20,21,23,25,26,27}"
DOCKER_CRON_SCHEDULE="${DOCKER_CRON_SCHEDULE:=0 4 * * *}"

sed -i "s~HYDRUSRV_CONTENT_DB_PATH=~HYDRUSRV_CONTENT_DB_PATH=$HYDRUSRV_CONTENT_DB_PATH~g" .env
sed -i "s~HYDRUS_SERVER_DB_PATH=~HYDRUS_SERVER_DB_PATH=$HYDRUS_SERVER_DB_PATH~g" .env
sed -i "s~HYDRUS_MASTER_DB_PATH=~HYDRUS_MASTER_DB_PATH=$HYDRUS_MASTER_DB_PATH~g" .env
sed -i "s~HYDRUS_MAPPINGS_DB_PATH=~HYDRUS_MAPPINGS_DB_PATH=$HYDRUS_MAPPINGS_DB_PATH~g" .env
sed -i "s~HYDRUS_TAG_REPOSITORY=2~HYDRUS_TAG_REPOSITORY=$HYDRUS_TAG_REPOSITORY~g" .env
sed -i "s~HYDRUS_FILE_REPOSITORY=3~HYDRUS_FILE_REPOSITORY=$HYDRUS_FILE_REPOSITORY~g" .env
sed -i "s~HYDRUS_SUPPORTED_MIME_TYPES=1,2,3,4,9,14,18,20,21,23,25,26,27~HYDRUS_SUPPORTED_MIME_TYPES=$HYDRUS_SUPPORTED_MIME_TYPES~g" .env

sed -i "s~DOCKER_CRON_SCHEDULE~$DOCKER_CRON_SCHEDULE~g" ./crontab

stop() {
  pkill supercronic
  sleep 1
}

trap "stop" SIGTERM

echo "waiting 60 seconds before running initial sync..."

sleep 60

node ./bin/sync

supercronic ./crontab &

wait $!
